### ISSUE TITLE
soname versioning

### ISSUE THREAD

**Issue by jeroen** [libgit2/libgit2#5670]
If the ABI is stable perhaps there is no need to keep bumping the soname? I couldn't find anything about the soname policy in the readme file. A good example is [libcurl soname policy](https://github.com/curl/curl/blob/master/docs/libcurl/ABI.md#soname-bumps) (which hasn't changed their soname since 2006).

It would be really nice if the soname was stable, so that distributions don't have to rebuild all reverse dependencies when they update libgit2.

**Comment by ethomson** [comment]
> If the ABI is stable perhaps there is no need to keep bumping the soname?

The ABI isn't stable.  This is an effort that we'll tackle in 2.0.  If we do that, we'll stabilize the soname.

**Comment by carterbox** [comment]
Is it time to reconsider how the soname is chosen for this project? I read through some of [this discussion about API stability](https://github.com/libgit2/libgit2/discussions/5882?sort=old), and I got the feeling that the developers of libgit2 may not be aware that the ABI and API of a library may be versioned separately.

> The ABI isn't stable.

For packagers, I would say it's more important that the soname accurately reflects ABI breaks than the ABI be stable.

On unix platforms, the linker uses the soname to decide whether two libraries are ABI compatible, but also multiple soversions of a library can coexist on a single machine. So it's much easier for packagers if the soname actually tracks with ABI-breaks because then downstream packages which are built against different API versions of libgit2 can coexist in the same environment because they can each link to their own ABI-compatible libgit.

I guess what I am proposing is that libgit2 transition from using a single version to separately versioning the API and ABI where the major version of the soname changes whenever there is an ABI breaking change (which may not conincide with a breaking API change).

For example, based on my cursory reading of the release notes, the ABI/API versions could have been as such:

API / release | ABI / soname | notes
-- | -- | -- |
1.0.0 | 1.0.0 |
1.0.1 | 1.0.1 |
1.1.0 | 1.1.0 |
1.1.1 | 1.1.1 |
1.2.0 | 1.2.0 |
1.3.0 | 1.3.0 |
1.4.0 | 2.0.0 | ABI breaks, but API is backwards compatible |
1.4.1 | 2.0.2 |
1.4.2 | 2.0.2 |
1.5.0 | 2.1.0 |
1.6.1 | 2.2.1 |
1.7.0 | 3.0.0 | ABI breaks, but API is backwards compatible |
1.8.0 | 4.0.0 | ABI breaks, but API is backwards compatible |

Using such a versioning scheme, packagers could ship just 4 artifacts: `libgit2.so.1.3`, `libgit2.so.2.2`, `libgit2.so.3.0`, `libgit2.so.4.0` all at the same time, and have an ABI-compatible version of the library for any downstream package compiled against `libgit2 >=1.0.0 (API)`

**Comment by ethomson** [comment]
@carterbox Yeah, we've probably ignored this issue for too long. We probably invest in getting this right.

**Comment by carterbox** [comment]
I offer to help write code, but I think the solution is more policy/procedure first than it is engineering. Though, there are some filename details when it comes to making linking consistent across windows/unix.

**Comment by ethomson** [comment]
We'll update the soname properly (I think) in v1.9.0, I've updated this in #6979.

One thing that I'll want to do is make sure that no distribution has tried to be helpful and fix our longstanding misunderstanding of libtool behavior. If somebody was shipping `libgit2-42.0.0.so` - which might make sense for them to have done - then we should make sure that we use a suitably large number here so that they can stop maintaining their own ABI / soname version scheme and just start using ours. I don't _think_ I've seen or heard of this, but I'll glance over some of the distributions before I make this update.

**Comment by ethomson** [comment]
Independent of this, I _would_ like to make the ABI more stable, but we're going to run a train on it when we add SHA256 support, so this seems like a post-v2.0 concern.
